cmd/thanos/query.go

```diff
@@ -215,30 +217,46 @@ func runQuery(
 	}
 
 	fileSDCache := cache.New()
-	dnsProvider := dns.NewProvider(
+	dnsStoreProvider := dns.NewProvider(
 		logger,
 		extprom.WrapRegistererWithPrefix("thanos_querier_store_apis_", reg),
 		dns.ResolverType(dnsSDResolver),
 	)
 
+	dnsRuleProvider := dns.NewProvider(
+		logger,
+		extprom.WrapRegistererWithPrefix("thanos_querier_rule_apis_", reg),
+		dns.ResolverType(dnsSDResolver),
+	)
+
 	var (
 		stores = query.NewStoreSet(
 			logger,
 			reg,
 			func() (specs []query.StoreSpec) {
 				// Add DNS resolved addresses from static flags and file SD.
-				for _, addr := range dnsProvider.Addresses() {
+				for _, addr := range dnsStoreProvider.Addresses() {
 					specs = append(specs, query.NewGRPCStoreSpec(addr))
 				}
 
 				specs = removeDuplicateStoreSpecs(logger, duplicatedStores, specs)
 
 				return specs
 			},
+			func() (specs []query.RuleSpec) {
+				for _, addr := range dnsRuleProvider.Addresses() {
+					specs = append(specs, query.NewGRPCStoreSpec(addr))
+				}
+
+				// NOTE(s-urbaniak): No need to remove duplicates, as rule apis are a subset of store apis.
```

Comment:
hmm, really?

Reply:
I thought this is fine by design, just you need to make sure RulesAPI implements StoreAPI as well?

Reply:
Correct, currently RulesAPI is a strict subset of StoreAPI, and we currently rely on that invariant.